deps/rabbitmq_mqtt/src/rabbit_mqtt_retained_msg_store_dets.erl

```diff
@@ -2,68 +2,308 @@
+  ok = dets:insert(NodeTable, {RootId, 0, false}),
+
+  #store_state{node_table = NodeTable,
+               edge_table = EdgeTable,
+               msg_table = MsgTable,
+               root_id = RootId,
+               dir = Dir,
+               vhost = VHost}.
 
 -spec recover(file:name_all(), rabbit_types:vhost()) ->
-    {ok, store_state(), rabbit_mqtt_retained_msg_store:expire()} |
-    {error, uninitialized}.
+               {ok, store_state(), rabbit_mqtt_retained_msg_store:expire()} |
+               {error, uninitialized}.
 recover(Dir, VHost) ->
-    case open_table(Dir, VHost) of
-        {ok, TabName}  ->
-            {ok,
-             #store_state{table = TabName},
-             rabbit_mqtt_retained_msg_store:expire(dets, TabName)};
-        {error, Reason} ->
-            ?LOG_ERROR("~s failed to open table: ~p", [?MODULE, Reason]),
-            {error, uninitialized}
-    end.
+  try
+    {ok, MsgTable} = open_table(Dir, VHost, <<"msgs">>, bag),
+    Expire = rabbit_mqtt_retained_msg_store:expire(dets, MsgTable),
+    {ok, NodeTable} = open_table(Dir, VHost, <<"nodes">>, set),
+    {ok, EdgeTable} = open_table(Dir, VHost, <<"edges">>, set),
 
--spec insert(topic(), mqtt_msg(), store_state()) -> ok.
-insert(Topic, Msg, #store_state{table = T}) ->
-  ok = dets:insert(T, #retained_message{topic = Topic, mqtt_msg = Msg}).
-
--spec lookup(topic(), store_state()) ->
-    mqtt_msg() | mqtt_msg_v0() | undefined.
-lookup(Topic, #store_state{table = T}) ->
-  case dets:lookup(T, Topic) of
-    []      -> undefined;
-    [#retained_message{mqtt_msg = Msg}] -> Msg
+    RootId =
+      case find_root_node(NodeTable, EdgeTable) of
+        {ok, Id} ->
+          ?LOG_INFO("Recovered existing RootId: ~p", [Id]),
+          Id;
+        error ->
+          NewId = make_node_id(),
+          ?LOG_INFO("Creating new RootId: ~p", [NewId]),
+          ok = dets:insert(NodeTable, {NewId, 0, false}),
+          NewId
+      end,
+
+    State =
+      #store_state{node_table = NodeTable,
+                   edge_table = EdgeTable,
+                   msg_table = MsgTable,
+                   root_id = RootId,
+                   dir = Dir,
+                   vhost = VHost},
+    {ok, State, Expire}
+  catch
+    error:Reason ->
+      ?LOG_ERROR("Failed to recover MQTT retained message store: ~p", [Reason]),
+      {error, uninitialized}
   end.
 
+-spec insert(topic(), mqtt_msg(), store_state()) -> ok.
+insert(Topic, Msg, #store_state{} = State) ->
+  Words = split_topic(Topic),
+  Insert =
+    fun() ->
+       NodeId = follow_or_create_path(Words, State),
+       % Mark node as topic end and store message
+       update_node(NodeId, true, State),
+       % Replace any existing message for this topic
+       dets:match_delete(State#store_state.msg_table, {NodeId, Topic, '_'}),
+       dets:insert(State#store_state.msg_table, {NodeId, Topic, Msg})
+    end,
+  Insert(),
+  %? TODO: Wrap in transaction since we're modifying multiple tables
+  ok.
```

Comment:
Should this operation be worked on to be atomic?

Reply:
I'll keep that in mind. Would using Khepri help to solve #8096?

Reply:
Yes, by definition Khepri is replicated. In fact, it might be "too replicated" because replicating to all nodes is excessive (for this kind of data).